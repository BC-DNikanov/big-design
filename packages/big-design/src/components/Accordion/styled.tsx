import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

export const StyledTitle = styled.div`
  color: ${({ theme }) => theme.colors.primary};
  cursor: pointer;
`;

StyledTitle.defaultProps = { theme: defaultTheme };
