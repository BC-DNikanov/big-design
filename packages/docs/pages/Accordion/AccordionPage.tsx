import { Accordion, H0, H1, Link, Text } from '@bigcommerce/big-design';
import React from 'react';

import { Code, CodePreview } from '../../components';
import { AccordionPropTable } from '../../PropTables';

export default () => (
  <>
    <H0>Accordion</H0>

    <Text>
      An <Code primary>Accordion</Code> component allows to show/hide a piece of data which is not relevant for users in
      a majority of use cases in order to save screen space.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [isExpanded, setIsExpanded] = React.useState(false);
        const handleToggle = () => setIsExpanded(!isExpanded);

        return (
          <Accordion title={isExpanded ? 'Show less' : 'Show more'} onToggle={handleToggle}>
            <Text>
              Ea tempor sunt amet labore proident dolor proident commodo in exercitation ea nulla sunt pariatur. Nulla
              sunt ipsum do eu consectetur exercitation occaecat labore aliqua. Aute elit occaecat esse ea fugiat esse.
              Reprehenderit sunt ea ea mollit commodo tempor amet fugiat.
            </Text>
          </Accordion>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>

    <H1>API</H1>
    <AccordionPropTable />
  </>
);
