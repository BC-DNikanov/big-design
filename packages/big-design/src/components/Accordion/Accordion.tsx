import { ExpandLessIcon, ExpandMoreIcon } from '@bigcommerce/big-design-icons';
import React, { useState } from 'react';

import { Flex } from '../Flex';

import { StyledTitle } from './styled';

export interface AccordionProps {
  title: string;
  onToggle?(): void;
}

export const Accordion: React.FC<AccordionProps> = ({ children, title, onToggle }) => {
  const [isOpened, setIsOpened] = useState(false);

  const toggleAccordion = () => setIsOpened(!isOpened);

  const handleTitleClick = () => {
    toggleAccordion();
    if (onToggle) {
      onToggle();
    }
  };

  return (
    <>
      <StyledTitle onClick={handleTitleClick}>
        {title}
        {isOpened ? <ExpandLessIcon /> : <ExpandMoreIcon />}
      </StyledTitle>
      <Flex display={isOpened ? 'flex' : 'none'} flexWrap="wrap">
        {children}
      </Flex>
    </>
  );
};

Accordion.displayName = 'Accordion';
