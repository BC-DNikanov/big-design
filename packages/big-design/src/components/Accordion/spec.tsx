import { fireEvent, render } from '@test/utils';
import 'jest-styled-components';
import React from 'react';

import { Accordion } from './Accordion';

test('render Accordion in default state', () => {
  const { container } = render(<Accordion title="title" />);

  expect(container).toMatchSnapshot();
  expect(container.lastChild).toHaveStyle('display: none');
});

test('switch Accordion to expanded state', () => {
  const { container } = render(<Accordion title="title" />);
  const trigger = container.firstChild as HTMLDivElement;

  fireEvent.click(trigger);

  expect(container).toMatchSnapshot();
  expect(container.lastChild).toHaveStyle('display: flex');
});

test('toggle content visibility', () => {
  const { container } = render(<Accordion title="title" />);
  const trigger = container.firstChild as HTMLDivElement;

  fireEvent.click(trigger);

  expect(container).toMatchSnapshot();
  expect(container.lastChild).toHaveStyle('display: flex');

  fireEvent.click(trigger);

  expect(container).toMatchSnapshot();
  expect(container.lastChild).toHaveStyle('display: none');
});

test('toggle title', () => {
  let isExpanded = false;
  const setIsExpanded = jest.fn(() => (isExpanded = !isExpanded));
  const handleToggle = jest.fn(() => setIsExpanded());
  const { container, rerender } = render(
    <Accordion title={isExpanded ? 'Show less' : 'Show more'} onToggle={handleToggle} />,
  );
  const trigger = container.firstChild as HTMLDivElement;

  fireEvent.click(trigger);

  rerender(<Accordion title={isExpanded ? 'Show less' : 'Show more'} onToggle={handleToggle} />);

  expect(container.firstChild).toMatchSnapshot();
  expect(handleToggle).toHaveBeenCalled();
  expect(isExpanded).toBe(true);
  expect(trigger.textContent).toBe('Show less');

  fireEvent.click(trigger);

  rerender(<Accordion title={isExpanded ? 'Show less' : 'Show more'} onToggle={handleToggle} />);

  expect(container.firstChild).toMatchSnapshot();
  expect(handleToggle).toHaveBeenCalled();
  expect(isExpanded).toBe(false);
  expect(trigger.textContent).toBe('Show more');
});
