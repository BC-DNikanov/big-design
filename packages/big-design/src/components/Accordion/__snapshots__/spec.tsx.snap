// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Accordion in default state 1`] = `
.c1 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c2 {
  display: none;
  box-sizing: border-box;
}

.c3 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: none;
}

.c0 {
  color: #3C64F4;
  cursor: pointer;
}

@media (min-width:0px) {
  .c3 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c3 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div>
  <div
    class="c0"
  >
    title
    <svg
      aria-labelledby="bd-icon-1"
      class="c1"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M24 24H0V0h24v24z"
        fill="none"
        opacity="0.87"
      />
      <path
        d="M15.88 9.29L12 13.17 8.12 9.29a.996.996 0 10-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 000-1.41c-.39-.38-1.03-.39-1.42 0z"
      />
    </svg>
  </div>
  <div
    class="c2 c3"
    display="none"
  />
</div>
`;

exports[`switch Accordion to expanded state 1`] = `
.c1 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
}

.c3 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  color: #3C64F4;
  cursor: pointer;
}

@media (min-width:0px) {
  .c3 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c3 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div>
  <div
    class="c0"
  >
    title
    <svg
      aria-labelledby="bd-icon-2"
      class="c1"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M11.29 8.71L6.7 13.3a.996.996 0 101.41 1.41L12 10.83l3.88 3.88a.996.996 0 101.41-1.41L12.7 8.71a.996.996 0 00-1.41 0z"
      />
    </svg>
  </div>
  <div
    class="c2 c3"
    display="flex"
  />
</div>
`;

exports[`toggle content visibility 1`] = `
.c1 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
}

.c3 {
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  color: #3C64F4;
  cursor: pointer;
}

@media (min-width:0px) {
  .c3 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:720px) {
  .c3 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div>
  <div
    class="c0"
  >
    title
    <svg
      aria-labelledby="bd-icon-2"
      class="c1"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M11.29 8.71L6.7 13.3a.996.996 0 101.41 1.41L12 10.83l3.88 3.88a.996.996 0 101.41-1.41L12.7 8.71a.996.996 0 00-1.41 0z"
      />
    </svg>
  </div>
  <div
    class="c2 c3"
    display="flex"
  />
</div>
`;

exports[`toggle content visibility 2`] = `
<div>
  <div
    class="styled__StyledTitle-sc-10nkrcr-0 cFktma"
  >
    title
    .c0 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

<svg
      aria-labelledby="bd-icon-3"
      class="c0"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M24 24H0V0h24v24z"
        fill="none"
        opacity="0.87"
      />
      <path
        d="M15.88 9.29L12 13.17 8.12 9.29a.996.996 0 10-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 000-1.41c-.39-.38-1.03-.39-1.42 0z"
      />
    </svg>
  </div>
  <div
    class="styled__StyledBox-sj5f1m-0 jBQGQm styled__StyledFlex-hcvk8l-0 entofW"
    display="none"
  />
</div>
`;

exports[`toggle title 1`] = `
.c1 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c0 {
  color: #3C64F4;
  cursor: pointer;
}

<div
  class="c0"
>
  Show less
  <svg
    aria-labelledby="bd-icon-2"
    class="c1"
    fill="currentColor"
    height="24"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M11.29 8.71L6.7 13.3a.996.996 0 101.41 1.41L12 10.83l3.88 3.88a.996.996 0 101.41-1.41L12.7 8.71a.996.996 0 00-1.41 0z"
    />
  </svg>
</div>
`;

exports[`toggle title 2`] = `
<div
  class="styled__StyledTitle-sc-10nkrcr-0 cFktma"
>
  Show more
  .c0 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

<svg
    aria-labelledby="bd-icon-2"
    class="c0"
    fill="currentColor"
    height="24"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M24 24H0V0h24v24z"
      fill="none"
      opacity="0.87"
    />
    <path
      d="M15.88 9.29L12 13.17 8.12 9.29a.996.996 0 10-1.41 1.41l4.59 4.59c.39.39 1.02.39 1.41 0l4.59-4.59a.996.996 0 000-1.41c-.39-.38-1.03-.39-1.42 0z"
    />
  </svg>
</div>
`;
